internal/lsp/converters.go

```diff
@@ -188,23 +192,60 @@ func fileNameToDocumentUri(fileName string) lsproto.DocumentUri {
 	return lsproto.DocumentUri("file://" + fileName)
 }
 
-func lineAndCharacterToPosition(lineAndCharacter lsproto.Position, lineMap []core.TextPos) int {
-	line := int(lineAndCharacter.Line)
-	offset := int(lineAndCharacter.Character)
+func lineAndCharacterToPosition(scriptInfo *project.ScriptInfo, lineAndCharacter lsproto.Position) core.TextPos {
+	// UTF-8/16 0-indexed line and character to UTF-8 offset
 
-	if line < 0 || line >= len(lineMap) {
-		panic(fmt.Sprintf("bad line number. Line: %d, lineMap length: %d", line, len(lineMap)))
+	lineMap := scriptInfo.LineMapLSP()
+
+	line := core.TextPos(lineAndCharacter.Line)
+	char := core.TextPos(lineAndCharacter.Character)
+
+	if line < 0 || int(line) >= len(lineMap.LineStarts) {
+		panic(fmt.Sprintf("bad line number. Line: %d, lineMap length: %d", line, len(lineMap.LineStarts)))
+	}
+
+	start := lineMap.LineStarts[line]
+	if lineMap.AsciiOnly {
+		return start + char
 	}
 
-	res := int(lineMap[line]) + offset
-	if line < len(lineMap)-1 && res >= int(lineMap[line+1]) {
-		panic("resulting position is out of bounds")
+	var utf8Char core.TextPos
+	var utf16Char core.TextPos
+
+	for i, r := range scriptInfo.Text()[start:] {
+		u16Len := core.TextPos(utf16.RuneLen(r))
+		if utf16Char+u16Len > char {
+			break
+		}
+		utf16Char += u16Len
+		utf8Char = core.TextPos(i + utf8.RuneLen(r))
 	}
-	return res
+
+	return start + utf8Char
 }
 
-func positionToLineAndCharacter(position int, lineMap []core.TextPos) lsproto.Position {
-	line, character := core.PositionToLineAndCharacter(position, lineMap)
+func positionToLineAndCharacter(scriptInfo *project.ScriptInfo, position core.TextPos) lsproto.Position {
+	// UTF-8 offset to UTF-8/16 0-indexed line and character
+
+	lineMap := scriptInfo.LineMapLSP()
```

Comment:
Does this ever get cached anywhere? Is the work being done on every call?

Reply:
Yes, scriptInfo caches this.

Reply:
Can you update the source file's line map to be the same if the file contains only ASCII? That way the two will be deduplicated.

Reply:
Or possibly just track whether a non CR/LF line ending was encountered here and do it then.

Reply:
By that point we'll have constructed the whole thing, so I'm not totally sure if it's helpful, but I guess we could save a little memory sometimes... If we've even requested the line map for other reasons, which we might not have at all.